README.md
# Component Gauge

Simple React Component to create Gauge.

![npm bundle size](https://img.shields.io/bundlephobia/minzip/components-gauge?style=for-the-badge)
![NPM](https://img.shields.io/npm/l/components-gauge?style=for-the-badge)
![npm](https://img.shields.io/npm/v/components-gauge?style=for-the-badge)
![npm](https://img.shields.io/npm/dw/components-gauge?style=for-the-badge)

Demo: <https://components-gauge.vercel.app/>

```bash
# npm
npm i components-gauge --save

# yarn
yarn add components-gauge

# pnpm
pnpm i components-gauge
```

### Usage

```jsx
import React, { Component } from "react";
import Gauge from "components-gauge";

export default function App() {
  return (
    <Gauge
      animated: true,
      width: 200,
      height: 200,
      baseStroke: 2,
      filledStroke: 2.5,
      baseColor: "#eee",
      filledColor: "#666",
      textColor: "#000",
      radius: 40,
      showValue: true,
    />
  );
}
```

## Props

| Prop                 | Default              | Description                                                                                                      |
| -------------------- | -------------------- | ---------------------------------------------------------------------------------------------------------------- |
| animated             | true                 | Animation                                                                                                        |
| width                | 200                  | width of chart 200px                                                                                             |
| height               | 200                  | height of chart 200px                                                                                            |
| baseStroke           | 2                    |                                                                                                                  |
| filledStroke         | 2.5                  |                                                                                                                  |
| baseColor            | #eee                 |                                                                                                                  |
| filledColor          | #666                 |                                                                                                                  |
| textColor            | #999                 |                                                                                                                  |
| radius               | 40                   |                                                                                                                  |
| showValue            | false                |                                                                                                                  |
| time                 | 10000(10s)           |                                                                                                                  |
| onEnd                | Empty                | Event on end                                                                                                     |

## License

MIT Licensed.


changelog.md


## [0.0.41](https://github.com/akbaruddin/components-gauge/compare/0.0.40...0.0.41) (2023-01-11)


### Bug Fixes

* broken cammand ([5c2e4dc](https://github.com/akbaruddin/components-gauge/commit/5c2e4dcd120192e2c0a8b762a719b85ec961d647))

## [0.0.40](https://github.com/akbaruddin/components-gauge/compare/0.0.35...0.0.40) (2022-10-11)

## [0.0.39](https://github.com/akbaruddin/components-gauge/compare/0.0.35...0.0.39) (2022-10-11)

## [0.0.38](https://github.com/akbaruddin/components-gauge/compare/0.0.35...0.0.38) (2022-10-11)

## [0.0.37](https://github.com/akbaruddin/components-gauge/compare/0.0.35...0.0.37) (2022-10-11)

## [0.0.36](https://github.com/akbaruddin/components-gauge/compare/0.0.35...0.0.36) (2022-10-11)

## [0.0.35](https://github.com/akbaruddin/components-gauge/compare/0.0.34...0.0.35) (2022-10-10)

## 0.0.34 (2022-10-10)


### Bug Fixes

* **version:** relase check ([cca5aaa](https://github.com/akbaruddin/components-guage/commit/cca5aaad0a1709c3e836cb2e064a7ab6c406aa47))

src/components/Gauge.js
import { useEffect, useReducer, useRef, useState } from "react";
import Animation from "./animation";
import { normalize } from "./utils";
import pathString from "./pathString";
import { getAngle, getValueInPercentage } from "./utils";
import useGaugeWindow from "./use-gauge-window";

const scaled = {
  transformOrigin: "center center",
  transform: "scale(2.5)",
};

const initState = {
  limit: 100,
  min: 0,
  oldValue: 0,
  value: 0,
  radius: 40,
  showValue: false,
  startAngle: -90,
  endAngle: -90.001,
  textClass: "value-text",
  valueLabelClass: null,
  baseClass: "dial",
  gaugeClass: "gauge",
  gaugeColor: () => null,
  viewBox: `0 0 100 100`,
  loaded: false,
  time: 10000,
};

const reducer = (state, action) => {
  if (action.type === "VALUE") {
    return {
      ...state,
      oldValue: state.value,
      value: action.payload,
    };
  }

  return state;
};

const Gauge = (props) => {
  const {
    animated = true,
    width = 200,
    height = 200,
    baseStroke = 2,
    filledStroke = 2.5,
    baseColor = "#eee",
    filledColor = "#666",
    textColor = "#999",
    radius = 40,
    label = (value) => Math.round(value * 100) / 100,
    time = 10000,
    showValue = false,
    min = 0,
    max = 100,
    gaugeColor = () => null,
    filledStrokeDasharray = "",
    start = -90,
    end = -90.001,
    textClass = "value-text",
    postfix = null,
    postfixClass = "post-fix",
    baseFill = "none",
    baseClass = "dial",
    filledClass = "value",
    onEnd = () => {},
    frame = 1,
    fullCircle = false,
  } = props;
  const { requestAnimationFrame } = useGaugeWindow();
  const [state, dispatch] = useReducer(reducer, {
    ...initState,
    radius,
    label,
    time,
    showValue,
    textColor,
    min,
    limit: max,
    gaugeColor,
    filledStrokeDasharray,
    startAngle: start < end ? end : start,
    endAngle: start < end ? start : end,
    textClass,
    postfix,
    postfixClass,
    baseFill,
    baseClass,
    filledClass,
  });
  const [text, setText] = useState(null);
  const gaugeContainer = useRef(null);
  const gaugeValueElem = useRef(null);
  const gaugeValuePath = useRef(null);
  const timer = useRef(null);
  const oldValueRef = useRef(0);
  const [base, setBase] = useState("");
  const [baseTop, setBaseTop] = useState("");

  const [gaugeDefaults] = useState({
    centerX: 50,
    centerY: 50,
  });

  const updateGauge = (theValue, frame) => {
    const val = getValueInPercentage(theValue, state.min, state.limit);
    const angle = getAngle(
      val,
      360 - Math.abs(state.startAngle - state.endAngle)
    );
    // this is because we are using arc greater than 180deg
    let flag = angle <= 180 ? 0 : 1;
    if (state.showValue) {
      setText(state.label.call(state, theValue));
    }
    setBaseTop(
      pathString(
        gaugeDefaults,
        state.radius,
        state.startAngle,
        angle + state.startAngle,
        flag
      )
    );
  };

  const setGaugeColor = (value, duration) => {
    const c = state.gaugeColor.call(state, value);
    const dur = duration * 1000;
    const pathTransition = "stroke " + dur + "ms ease";
    // const textTransition = "fill " + dur + "ms ease";

    gaugeValuePath.current.style.stroke = c;
    gaugeValuePath.current.style["-webkit-transition"] = pathTransition;
    gaugeValuePath.current.style["-moz-transition"] = pathTransition;
    gaugeValuePath.current.style.transition = pathTransition;
    // gaugeValueElem.current.style = [
    //   "fill: " + c,
    //   "-webkit-transition: " + textTransition,
    //   "-moz-transition: " + textTransition,
    //   "transition: " + textTransition,
    // ].join(";");
  };

  const setValue = (val) => {
    const newValue = normalize(val, state.min, state.limit);
    dispatch({ type: "VALUE", payload: newValue });
    if (state.gaugeColor) {
      setGaugeColor(newValue, 0);
    }
    updateGauge(newValue);
  };

  const setValueAnimated = (val, duration, callback = () => {}) => {
    const oldValue = oldValueRef.current;
    const newValue = normalize(val, state.min, state.limit);
    if (oldValue === newValue) {
      return;
    }

    if (state.gaugeColor) {
      setGaugeColor(newValue, duration);
    }
    Animation(
      {
        start: oldValue || 0,
        end: newValue,
        duration: duration || 1,
        step: (val, frame) => {
          updateGauge(val, frame);
        },
      },
      requestAnimationFrame
    );
    dispatch({ type: "VALUE", payload: newValue });
    oldValueRef.current = newValue;
  };

  useEffect(() => {
    setBaseTop(
      pathString(
        gaugeDefaults,
        state.radius,
        state.startAngle,
        state.startAngle
      )
    );

    const angle = getAngle(
      100,
      360 - Math.abs(state.startAngle - state.endAngle)
    );
    const flag = angle <= 180 ? 0 : 1;
    if (fullCircle) {
      setBase("M 50 40 A 10 10 0 1 1 49.999 40");
    } else {
      setBase(
        pathString(
          gaugeDefaults,
          state.radius,
          state.startAngle,
          state.endAngle,
          flag
        )
      );
    }

    const totalTime = state.time / 1000;
    let endTime = 0;
    timer.current = setInterval(() => {
      if (endTime >= totalTime) {
        clearInterval(timer.current);
        onEnd();
      } else {
        endTime += 1;
        let circleEnd = state.limit * (endTime / totalTime);
        if (fullCircle && circleEnd === state.limit) {
          circleEnd -= 0.001;
        }
        if (animated) {
          setValueAnimated(circleEnd, frame);
        } else {
          setValue(circleEnd);
        }
      }
    }, 1000);

    return () => {
      clearInterval(timer.current);
    };
  }, []);

  return (
    <div ref={gaugeContainer} style={{ width: width, height: height }}>
      <svg
        viewBox={state.viewBox}
        xmlns="http://www.w3.org/2000/svg"
        className={state.gaugeClass}
      >
        <path
          className={state.baseClass}
          fill={state.baseFill}
          stroke={baseColor}
          strokeWidth={baseStroke}
          d={base}
          style={fullCircle ? scaled : {}}
        />
        {state.showValue && (
          <g className="text-container">
            <text
              x={50}
              y={50}
              fill={state.textColor}
              className={state.textClass}
              fontSize="100%"
              fontFamily="sans-serif"
              fontWeight="normal"
              textAnchor="middle"
              alignmentBaseline="middle"
              dominantBaseline="central"
              ref={gaugeValueElem}
            >
              {text}
            </text>
            {state.postfix && (
              <text className={state.postfixClass}>{state.postfix}</text>
            )}
          </g>
        )}
        <path
          className={state.filledClass}
          fill="none"
          stroke={filledColor}
          strokeWidth={filledStroke}
          d={baseTop}
          ref={gaugeValuePath}
          strokeDasharray={state.filledStrokeDasharray}
          style={fullCircle ? scaled : {}}
        />
      </svg>
    </div>
  );
};

export default Gauge;
